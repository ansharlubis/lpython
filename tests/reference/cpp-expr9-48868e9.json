{
    "basename": "cpp-expr9-48868e9",
    "cmd": "lpython --no-color --show-cpp {infile}",
    "infile": "tests/expr9.py",
    "infile_hash": "1f02a7486b298ae9e74a163875a76f8fa7cc25d7f50547133dbbdfab",
    "outfile": null,
    "outfile_hash": null,
    "stdout": "cpp-expr9-48868e9.stdout",
    "stdout_hash": "90833fa180ea1a6d3227a9a485514b484f083d5766f6c8c20222b2bb",
    "stderr": null,
    "stderr_hash": null,
    "returncode": 0
}