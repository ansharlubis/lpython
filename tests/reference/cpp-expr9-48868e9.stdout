([1m[35mTranslationUnit[39m[0m 
   ([33mSymbolTable[39m 
      1 
      {
         main0: 
            ([1m[35mSubroutine[39m[0m 
               ([33mSymbolTable[39m 
                  6 
                  {
                     i: 
                        ([1m[35mVariable[39m[0m 
                           6 
                           i [1m[32m
                           Local[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     s: 
                        ([1m[35mVariable[39m[0m 
                           6 
                           s [1m[32m
                           Local[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mCharacter[39m[0m 1 -2 () []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.)
                  }) 
               main0 [] [
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 6 [33mi[39m) 
                  ([1m[35mFunctionCall[39m[0m 1 [33mtest_return_1[39m () [
                     (([1m[35mIntegerConstant[39m[0m [36m4[39m 
                        ([1m[35mInteger[39m[0m 4 [])))] 
                     ([1m[35mInteger[39m[0m 4 []) () ()) ()) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 6 [33ms[39m) 
                  ([1m[35mFunctionCall[39m[0m 1 [33mtest_return_2[39m () [
                     (([1m[35mIntegerConstant[39m[0m [36m4[39m 
                        ([1m[35mInteger[39m[0m 4 [])))] 
                     ([1m[35mCharacter[39m[0m 1 -2 () []) () ()) ()) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 6 [33mi[39m) 
                  ([1m[35mFunctionCall[39m[0m 1 [33mtest_return_3[39m () [
                     (([1m[35mIntegerConstant[39m[0m [36m4[39m 
                        ([1m[35mInteger[39m[0m 4 [])))] 
                     ([1m[35mInteger[39m[0m 4 []) () ()) ()) 
               ([1m[35mSubroutineCall[39m[0m 1 [33mtest_return_4[39m () [
                  (([1m[35mIntegerConstant[39m[0m [36m4[39m 
                     ([1m[35mInteger[39m[0m 4 [])))] ())] [1m[32m
               Source[39m[0m [1m[32m
               Public[39m[0m [1m[32m
               Implementation[39m[0m () .false. .false.), 
         test_return_1: 
            ([1m[35mFunction[39m[0m 
               ([33mSymbolTable[39m 
                  2 
                  {
                     _lpython_return_variable: 
                        ([1m[35mVariable[39m[0m 
                           2 
                           _lpython_return_variable [1m[32m
                           ReturnVar[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     a: 
                        ([1m[35mVariable[39m[0m 
                           2 
                           a [1m[32m
                           In[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     x: 
                        ([1m[35mVariable[39m[0m 
                           2 
                           x [1m[32m
                           Local[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.)
                  }) 
               test_return_1 [
               ([1m[35mVar[39m[0m 2 [33ma[39m)] [
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 2 [33mx[39m) ([1m[35mIntegerConstant[39m[0m [36m5[39m 
                  ([1m[35mInteger[39m[0m 4 [])) ()) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 2 [33m_lpython_return_variable[39m) 
                  ([1m[35mVar[39m[0m 2 [33mx[39m) ()) 
               ([1m[35mReturn[39m[0m)] 
               ([1m[35mVar[39m[0m 2 [33m_lpython_return_variable[39m) [1m[32m
               Source[39m[0m [1m[32m
               Public[39m[0m [1m[32m
               Implementation[39m[0m ()), 
         test_return_2: 
            ([1m[35mFunction[39m[0m 
               ([33mSymbolTable[39m 
                  3 
                  {
                     _lpython_return_variable: 
                        ([1m[35mVariable[39m[0m 
                           3 
                           _lpython_return_variable [1m[32m
                           ReturnVar[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mCharacter[39m[0m 1 -2 () []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     a: 
                        ([1m[35mVariable[39m[0m 
                           3 
                           a [1m[32m
                           In[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     x: 
                        ([1m[35mVariable[39m[0m 
                           3 
                           x [1m[32m
                           Local[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mCharacter[39m[0m 1 -2 () []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.)
                  }) 
               test_return_2 [
               ([1m[35mVar[39m[0m 3 [33ma[39m)] [
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 3 [33mx[39m) 
                  ([1m[35mStringConstant[39m[0m "test" 
                     ([1m[35mCharacter[39m[0m 1 4 () [])) ()) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 3 [33m_lpython_return_variable[39m) 
                  ([1m[35mVar[39m[0m 3 [33mx[39m) ()) 
               ([1m[35mReturn[39m[0m)] 
               ([1m[35mVar[39m[0m 3 [33m_lpython_return_variable[39m) [1m[32m
               Source[39m[0m [1m[32m
               Public[39m[0m [1m[32m
               Implementation[39m[0m ()), 
         test_return_3: 
            ([1m[35mFunction[39m[0m 
               ([33mSymbolTable[39m 
                  4 
                  {
                     _lpython_return_variable: 
                        ([1m[35mVariable[39m[0m 
                           4 
                           _lpython_return_variable [1m[32m
                           ReturnVar[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.), 
                     a: 
                        ([1m[35mVariable[39m[0m 
                           4 
                           a [1m[32m
                           In[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.)
                  }) 
               test_return_3 [
               ([1m[35mVar[39m[0m 4 [33ma[39m)] [
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 4 [33ma[39m) ([1m[35mIntegerConstant[39m[0m [36m3[39m 
                  ([1m[35mInteger[39m[0m 4 [])) ()) 
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 4 [33m_lpython_return_variable[39m) 
                  ([1m[35mVar[39m[0m 4 [33ma[39m) ()) 
               ([1m[35mReturn[39m[0m)] 
               ([1m[35mVar[39m[0m 4 [33m_lpython_return_variable[39m) [1m[32m
               Source[39m[0m [1m[32m
               Public[39m[0m [1m[32m
               Implementation[39m[0m ()), 
         test_return_4: 
            ([1m[35mSubroutine[39m[0m 
               ([33mSymbolTable[39m 
                  5 
                  {
                     a: 
                        ([1m[35mVariable[39m[0m 
                           5 
                           a [1m[32m
                           In[39m[0m () () [1m[32m
                           Default[39m[0m 
                           ([1m[35mInteger[39m[0m 4 []) [1m[32m
                           Source[39m[0m [1m[32m
                           Public[39m[0m [1m[32m
                           Required[39m[0m .false.)
                  }) 
               test_return_4 [
               ([1m[35mVar[39m[0m 5 [33ma[39m)] [
               ([1m[35m=[39m[0m 
                  ([1m[35mVar[39m[0m 5 [33ma[39m) ([1m[35mIntegerConstant[39m[0m [36m1[39m 
                  ([1m[35mInteger[39m[0m 4 [])) ()) 
               ([1m[35mReturn[39m[0m)] [1m[32m
               Source[39m[0m [1m[32m
               Public[39m[0m [1m[32m
               Implementation[39m[0m () .false. .false.)
      }) 
   [
   ([1m[35mSubroutineCall[39m[0m 1 [33mmain0[39m () [] ())])
#include <iostream>
#include <string>
#include <vector>
#include <cassert>
#include <cmath>
#include <complex>
#include <Kokkos_Core.hpp>
#include <lfortran_intrinsics.h>

template <typename T>
Kokkos::View<T*> from_std_vector(const std::vector<T> &v)
{
    Kokkos::View<T*> r("r", v.size());
    for (size_t i=0; i < v.size(); i++) {
        r(i) = v[i];
    }
    return r;
}

// Forward declarations
void _lpython_main_program();
void main0();
int32_t test_return_1(int a);
std::string test_return_2(int a);
int32_t test_return_3(int a);
void test_return_4(int a);
namespace {
}

// Implementations
void _lpython_main_program()
{
    main0();
}

void main0()
{
    int i;
    std::string s;
    i = test_return_1(4);
    s = test_return_2(4);
    i = test_return_3(4);
    test_return_4(4);
}

int32_t test_return_1(int a)
{
    int _lpython_return_variable;
    int x;
    x = 5;
    _lpython_return_variable = x;
    return _lpython_return_variable;
}

std::string test_return_2(int a)
{
    std::string _lpython_return_variable;
    std::string x;
    x = "test";
    _lpython_return_variable = x;
    return _lpython_return_variable;
}

int32_t test_return_3(int a)
{
    int _lpython_return_variable;
    a = 3;
    _lpython_return_variable = a;
    return _lpython_return_variable;
}

void test_return_4(int a)
{
    a = 1;
    return;
}

namespace {

void main2() {
    _lpython_main_program();
}

}
int main(int argc, char* argv[])
{
    Kokkos::initialize(argc, argv);
    main2();
    Kokkos::finalize();
    return 0;
}
